include/ginkgo/core/distributed/vector.hpp

```diff
@@ -0,0 +1,359 @@
+     * scaled by alpha. If it is a Dense row vector of values, then i-th column
+     * vector of b is scaled with the i-th element of alpha (the number of c
+     * @param b  a (multi-)vector of the same dimension as this
+     */
+    void sub_scaled(const LinOp* alpha, const LinOp* b);
+
+    /**
+     * Computes the column-wise dot product of this (multi-)vector and `b` using
+     * a global reduction.
+     *
+     * @param b  a (multi-)vector of same dimension as this
+     * @param result  a Dense row matrix, used to store the dot product
+     *                (the number of column in result must match the number
+     *                of columns of this)
+     */
+    void compute_dot(const LinOp* b, LinOp* result) const;
+
+    /**
+     * Computes the column-wise dot product of this (multi-)vector and `conj(b)`
+     * using a global reduction.
+     *
+     * @param b  a (multi-)vector of same dimension as this
+     * @param result  a Dense row matrix, used to store the dot product
+     *                (the number of column in result must match the number
+     *                of columns of this)
+     */
+    void compute_conj_dot(const LinOp* b, LinOp* result) const;
+
+    /**
+     * Computes the Euclidian (L^2) norm of this (multi-)vector using a global
+     * reduction.
+     *
+     * @param result  a Dense row matrix, used to store the norm
+     *                (the number of columns in result must match the number
+     *                of columns of this)
+     */
+    void compute_norm2(LinOp* result) const;
+
+    /**
+     * Computes the column-wise (L^1) norm of this (multi-)vector.
+     *
+     * @param result  a Dense row matrix, used to store the norm
+     *                (the number of columns in result must match the number
+     *                of columns of this)
+     */
+    void compute_norm1(LinOp* result) const;
+
+    /**
+     * Direct (read) access to the underlying local local_vector_type vectors.
+     *
+     * @return a constant pointer to the underlying local_vector_type vectors
+     */
+    const local_vector_type* get_const_local() const;
+
+    /*
+     * Direct (read/write) access to the underlying local_vector_type Dense
+     * vectors.
+     *
+     * @return a constant pointer to the underlying local_vector_type vectors
+     */
+    local_vector_type* get_local();
+
+protected:
+    /**
+     * Creates an empty distributed vector with a specified size
+     * @param exec  Executor associated with vector
+     * @param comm  Communicator associated with vector, the default is
+     *              MPI_COMM_WORLD
+     * @param partition  Partition of global rows
+     * @param global_size  Global size of the vector
+     * @param local_size  Processor-local size of the vector
+     * @param stride  Stride of the local vector.
+     */
+    Vector(std::shared_ptr<const Executor> exec, mpi::communicator comm,
+           dim<2> global_size, dim<2> local_size, size_type stride);
```

Comment:
One constructor I am missing is one that takes an existing Dense matrix and wraps it into a distributed vector by computing the global size via allreduce (wrapping it here would mean move-assignment)

Reply:
So `Vector(exec, comm, std::move(local_vec))` without any further parameters right? I also will add a constructor `Vector(exec, comm, global_size, local_size, ValueArray&&, stride)` similar to the dense one. (I need that one for the `create_real_view` which is needed for the dispatching...)

Reply:
I don't think you need the second one, because you can use the corresponding Dense constructor and pass the matrix to the first constructor. We probably need a version with and without global_size, and only compute it (via allreduce) if necessary.

I mean matrix::Dense* local_mtx and local_mtx->move_to(&member_local_matrix), because users will most likely not use references to dense matrices, but pointers.

Reply:
Yes, of course, don't know what I was thinking there. But should it perhaps be `std::unique_ptr<local_type>` to make it clear that we will move from that?

Reply:
That might unnecessarily restrict from the user's perspective: If they do their own memory management, they first need to release the pointer from their own structures and then wrap it in unique_ptr. This might be a data point to keep in mind if we decide to tackle #179 at some point.